gobblin-api/src/main/java/org/apache/gobblin/converter/Converter.java

```diff
@@ -120,16 +125,27 @@ public State getFinalState() {
   public RecordStreamWithMetadata<DO, SO> processStream(RecordStreamWithMetadata<DI, SI> inputStream,
       WorkUnitState workUnitState) throws SchemaConversionException {
     init(workUnitState);
-    SO outputSchema = convertSchema(inputStream.getSchema(), workUnitState);
+    this.outputGlobalMetadata =
+        new GlobalMetadata<SO>(convertSchema(inputStream.getGlobalMetadata().getSchema(), workUnitState));
     Flowable<StreamEntity<DO>> outputStream =
         inputStream.getRecordStream()
             .flatMap(in -> {
               if (in instanceof ControlMessage) {
+                ControlMessage out = (ControlMessage) in;
+                // update the output schema with the new input schema from the MetadataUpdateControlMessage
+                if (in instanceof MetadataUpdateControlMessage) {
+                  this.outputGlobalMetadata =
+                      new GlobalMetadata<SO>(convertSchema((SI)((MetadataUpdateControlMessage) in).getGlobalMetadata()
+                                      .getSchema(), workUnitState));
+                  out = new MetadataUpdateControlMessage<SO, DO>(this.outputGlobalMetadata);
+                }
+
                 getMessageHandler().handleMessage((ControlMessage) in);
```

Comment:
Should this receive `in` or `out`? If the decision is for `in`, then I would prefer this line be before the `MetadataUpdateControlMessage` block.

Reply:
Okay, moved it before. I think `in` should be the one to be handled so that the handler gets a chance to handle the control message without modifications. The handler can choose to do schema conversion if it needs it.